### ISSUE TITLE
Representing decision settings

### ISSUE THREAD

**Issue by krusnim** [JuliaDecisionMaking/Decisions.jl#35]
I brought this up a couple weeks ago, but it's still on my mind. In the draft documentation, I have the following:

> In all decision problems, we seek some distribution(s) of agent behavior. However, so far,
we've developed these distributions _ex nihilo_ for individual problems, making single
decisions one at a time based on input ("online"). Often, we actually decide on higher-level
parameters or hyperparameters for a decision making process ("offline"). These more complex
decisions might involve repeated interactions with the environment, a human-defined
objective, or interactions with other optimizing agents.

> A `DecisionSetting` represents some sort of interaction between a decision network
([`DecisionNetwork`](@ref)) and one or more solvers ([`DecisionAlgorithm`](@ref)s) that
produces conditional distributions for the network's decision nodes (that is, any nodes for
which the network itself doesn't supply defintions). Once a setting is defined, we can
finally `decide` to make a decision. We might `decide` online or
offline, involving a single agent or across many agents, with or without persistent
parameters and hyperparameters, and so on. In the simplest (online) case, this interaction
just queries the algorithm for an output whenever a decision node is reached. 

**BLUF**: I want users to specify exactly how and when `DecisionAlgorithm`s interact with a
network (and each other) during the optimization process. It's the source of far more
confusion than "what DDN is this paper operating with / what specifically does the policy
get to know," addressing instead "what exactly does the algorithm get to use to make that
policy."  

But at the very least, we need a system like `DecisionMaking.jl`'s explicit
`DMAlgorithmVariables` and `DMAlgorithmHyperparameters`. IMO, we should leave the door open
for generalization to other settings than "online" / "offline parameter opt" / "offline
hyperparameter opt" while we're at it, even if we don't actually get around to defining
them.

## A possible setup
To me, the relevant functions are:

* `simulate(fn, ::DecisionNetwork, behavior::NamedTuple)`: This is the same `simulate` as
  always: run the (D)DN until `fn` says we're done or we reach a `Terminal()` state.

* `<x>Setting(::DecisionNetwork, ::DecisionAlgorithm...)`: E.g.,
`OnlineSetting(::DecisionNetwork, ::DecisionAlgorithm)`. This defines a particular decision
setting for the algorithm and network to interact in, per the documentation.

* `decide(fn, ::DecisionSetting)`: This specializes based on the `DecisionSetting` and
actually carries out the interaction between the environment and the algorithm. It probably
uses `simulate`, maybe iteratively. It could look like an online setup, an offline training
loop, a CT/DE setup, a hyperparameter sweep, etc. It can call various functions on the
`DecisionAlgorithm` (think `update_params`, etc); the functions it expects to be defined on
the algorithm depend on the setting, but in all cases the actual updates are performed by
the `DecisionAlgorithm`; the `DecisionSetting` just specifies when the algorithm actually
gets control and what information it gets. In an online setting this is pretty much
synonymous with `simulate`. `fn` has a similar interpretation as in `simulate`; it tells
`decide` when to stop, and maybe does other stuff. Its input, and how frequently it is
called, depends on the setting; it could be once per DNN step, once per episode, once per
hyperparameter setting, etc. Final output is a `NamedTuple` of `ConditionalDists` for all
decision nodes in the original network; i.e., a solution to the decision problem.

* `evaluate(::DecisionObjective, ::DecisionSetting)`: This specifies some sort of final,
bottom-line metric for a decision. `::DecisionObjective`s might be used elsewhere to define
a `DecisionSetting` (e.g., to define a training loop), and there they might be usable by a
`DecisionAlgorithm`, but algorithms _absolutely 100% never get to see the output of
`evaluate`_. Probably, but not necessarily, it uses `decide`. We could be evaluating...
  * the solution produced by `decide`, or 
  * some artifact produced along the way (like a training curve), or 
  * some meta-quantity about the decision process (like power usage, to borrow a
`DecisionMaking` example), or 
  * something only tangentially related to the setting at all (like how long it takes to
simulate the fully implemented DN).

I had at one point considered using a `DecisionGraph` to underlie `DecisionSettings`
(treating them like meta-DNs), but I finally decided that's just too confusing, so for now
`DecisionSettings` have their own (shallow) type hierarchy with no type parameters.

Anyway, IMO, naming particular decision settings is necessary to prevent what @kylewray
calls "shell games," and also just very handy to be able to do, even if we can only get the
most relevant ones. I'd appreciate input on this point, though, because it's pretty colored
by my adversarial multiagent background (where there's constant confusion about what, when,
and how agents get to know about other agents' optimization processes.)
